File: realtube-go/internal/service/score_worker.go
package service

import (
	"context"
	"log"
	"sync"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
)

// ScoreWorker listens for PostgreSQL NOTIFY on the 'vote_changes' channel
// and batches score recalculations. If 50 votes hit video X in 5 seconds,
// it recalculates once (infrastructure-design.md §14).
type ScoreWorker struct {
	pool     *pgxpool.Pool
	scoreSvc *ScoreService
	cache    *CacheService
	batchMs  time.Duration

	mu      sync.Mutex
	pending map[string]struct{} // video IDs waiting for recalculation
}

// NewScoreWorker creates a score recalculation worker.
func NewScoreWorker(pool *pgxpool.Pool, scoreSvc *ScoreService, cache *CacheService) *ScoreWorker {
	return &ScoreWorker{
		pool:     pool,
		scoreSvc: scoreSvc,
		cache:    cache,
		batchMs:  5 * time.Second,
		pending:  make(map[string]struct{}),
	}
}

// Start begins listening for vote_changes notifications and processing batches.
func (w *ScoreWorker) Start(ctx context.Context) {
	log.Printf("score-worker: starting (batch window=%s)", w.batchMs)

	for {
		if err := w.listenLoop(ctx); err != nil {
			if ctx.Err() != nil {
				log.Println("score-worker: stopping (context cancelled)")
				return
			}
			log.Printf("score-worker: listen error, reconnecting in 5s: %v", err)
			select {
			case <-time.After(5 * time.Second):
			case <-ctx.Done():
				log.Println("score-worker: stopping (context cancelled)")
				return
			}
		}
	}
}

// listenLoop acquires a dedicated connection, LISTENs on vote_changes,
// and processes notifications in batched windows.
func (w *ScoreWorker) listenLoop(ctx context.Context) error {
	conn, err := w.pool.Acquire(ctx)
	if err != nil {
		return err
	}
	defer conn.Release()

	_, err = conn.Exec(ctx, "LISTEN vote_changes")
	if err != nil {
		return err
	}
	log.Println("score-worker: listening on vote_changes")

	// Start the batch flush goroutine
	flushCtx, flushCancel := context.WithCancel(ctx)
	defer flushCancel()
	go w.flushLoop(flushCtx)

	for {
		notification, err := conn.Conn().WaitForNotification(ctx)
		if err != nil {
			return err
		}

		videoID := notification.Payload
		if videoID == "" {
			continue
		}

		w.mu.Lock()
		w.pending[videoID] = struct{}{}
		w.mu.Unlock()
	}
}

// flushLoop periodically drains the pending set and recalculates scores.
func (w *ScoreWorker) flushLoop(ctx context.Context) {
	ticker := time.NewTicker(w.batchMs)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			w.flush(ctx)
		case <-ctx.Done():
			// Final flush before exit
			w.flush(context.Background())
			return
		}
	}
}

// flush drains the pending set and recalculates each video's score.
func (w *ScoreWorker) flush(ctx context.Context) {
	w.mu.Lock()
	if len(w.pending) == 0 {
		w.mu.Unlock()
		return
	}

	// Swap out the pending map
	batch := w.pending
	w.pending = make(map[string]struct{})
	w.mu.Unlock()

	recalculated := 0
	for videoID := range batch {
		if err := w.scoreSvc.RecalculateVideoScore(ctx, videoID); err != nil {
			log.Printf("score-worker: recalculate error for %s: %v", videoID, err)
			continue
		}

		// Invalidate Redis cache so next read gets fresh data
		if w.cache != nil {
			if err := w.cache.InvalidateVideo(ctx, videoID); err != nil {
				log.Printf("score-worker: cache invalidate error for %s: %v", videoID, err)
			}
		}

		recalculated++
	}

	if recalculated > 0 {
		log.Printf("score-worker: batch complete — %d videos recalculated (from %d notifications)",
			recalculated, len(batch))
	}
}

File: realtube-python/app/services/score_worker.py
"""Async score recalculation worker using PostgreSQL LISTEN/NOTIFY.

Listens on the 'vote_changes' channel for video IDs, batches them over
a 5-second window, deduplicates, and recalculates once per unique video ID
(infrastructure-design.md §14).
"""

import asyncio
import logging
import time

import asyncpg

from app.services.score_service import recalculate_video_score

logger = logging.getLogger(__name__)

BATCH_WINDOW_SECONDS = 5


async def run(
    pool: asyncpg.Pool,
    cache_service,
    batch_window: float = BATCH_WINDOW_SECONDS,
) -> None:
    """Run the score worker loop forever.

    Args:
        pool: Database connection pool.
        cache_service: CacheService instance for Redis invalidation.
        batch_window: Seconds to collect notifications before flushing.
    """
    logger.info("score-worker: starting (batch window=%.0fs)", batch_window)

    while True:
        try:
            await _listen_loop(pool, cache_service, batch_window)
        except asyncio.CancelledError:
            logger.info("score-worker: stopping (cancelled)")
            return
        except Exception:
            logger.exception("score-worker: listen error, reconnecting in 5s")
            try:
                await asyncio.sleep(5)
            except asyncio.CancelledError:
                logger.info("score-worker: stopping (cancelled)")
                return


async def _listen_loop(
    pool: asyncpg.Pool,
    cache_service,
    batch_window: float,
) -> None:
    """Acquire a dedicated connection, LISTEN, and batch notifications."""
    conn = await pool.acquire()
    try:
        pending: set[str] = set()

        def _on_notification(conn, pid, channel, payload):
            if payload:
                pending.add(payload)

        await conn.add_listener("vote_changes", _on_notification)
        logger.info("score-worker: listening on vote_changes")

        try:
            while True:
                await asyncio.sleep(batch_window)

                if not pending:
                    continue

                # Swap out the pending set
                batch = pending.copy()
                pending.clear()

                await _flush(pool, cache_service, batch)
        finally:
            await conn.remove_listener("vote_changes", _on_notification)
    finally:
        await pool.release(conn)


async def _flush(
    pool: asyncpg.Pool,
    cache_service,
    batch: set[str],
) -> None:
    """Recalculate scores for all video IDs in the batch."""
    start = time.time()
    recalculated = 0

    for video_id in batch:
        try:
            await recalculate_video_score(pool, video_id)

            # Invalidate Redis cache
            if cache_service is not None:
                try:
                    await cache_service.invalidate_video(video_id)
                except Exception:
                    logger.warning(
                        "score-worker: cache invalidate error for %s",
                        video_id,
                        exc_info=True,
                    )

            recalculated += 1
        except Exception:
            logger.exception(
                "score-worker: recalculate error for %s", video_id
            )

    elapsed = time.time() - start
    if recalculated > 0:
        logger.info(
            "score-worker: batch complete — %d videos recalculated "
            "(from %d notifications, %.0fms)",
            recalculated,
            len(batch),
            elapsed * 1000,
        )
